bin/propolis-server/src/lib/vm/state_driver.rs

```diff
@@ -108,43 +139,98 @@ where
         self.api_state_tx.borrow().state
     }
 
-    /// Publishes the supplied externally-visible instance state to the external
-    /// instance state channel.
-    fn set_instance_state(&mut self, state: ApiInstanceState) {
-        let old = self.api_state_tx.borrow().clone();
+    /// Retrieves the most recently published migration state from the external
+    /// migration state channel.
+    ///
+    /// This function does not return the borrowed monitor, so the state may
+    /// change again as soon as this function returns.
+    fn get_migration_status(&self) -> ApiMigrateStatusResponse {
+        self.api_state_tx.borrow().migration.clone()
+    }
+
+    /// Sets the published instance and/or migration state and increases the
+    /// state generation number.
+    ///
+    /// # Panics
+    ///
+    /// Panics if both `instance_state` and `migration_state` are `None`.
```

Comment:
this is a bit of a bummer, I kind of wonder if we should instead have an 
```rust
enum PublishedState {
    Instance(ApiInstanceState),
    Migration(PublishedMigrationState),
    Both(ApiInstanceState, PublishedMigrationState),
}
```
or similar? On the other hand, if that's too annoying to work with, I don't think it's a huge deal...

Reply:
This was bothering me too on self-review. I'll have a fix in my next push.